### PR TITLE
Implmeneted data split for kafka message

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- Data split is implemented to publish Kafka message
- Key is not implemented for deciding which partition the message is sent to
- Header includes:
  - checksum for the data consistency
  - the number of total segments for a message
  - segment index

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by yoomee1313** [comment]
Could you please add some comments in TestKafka_split and TestKafka_makeProducerMessage function? It would be easier to understand the tests with more comments!

**Comment by jeongkyun-oh** [comment]
> Could you please add some comments in TestKafka_split and TestKafka_makeProducerMessage function? It would be easier to understand the tests with more comments!

Alright. I will add some comments.

File: datasync/chaindatafetcher/kafka/config.go [hunk 1]
```diff
@@ -40,6 +40,7 @@ const (
 	DefaultTopicResourceName    = "en-0"
 	DefaultMaxMessageBytes      = 1000000
 	DefaultRequiredAcks         = 1
+	DefaultSegmentSize          = 1000000 // 1 MB
```
Comment (kjhman21) [inline datasync/chaindatafetcher/kafka/config.go:43]: 1MB = 1024 * 1024 Bytes, isn't it?

**Comment by winnie-byun** [comment]
I have some questions and comments.
1. Why splitting is required in Kafka? What are the benefits?
2. I think publishing segments in parallel can enhance speed.
3. A data is sent by calling `publish()` one time, and is received by calling several `subscribe()`. How do we know if the received data is a splitted one? After receiving, the segments need to be combined into one.
  It would be nice to specify a protocol including data ID and segment number. Also, one `publish()` call should be correspond to one `subscribe()`.

File: datasync/chaindatafetcher/kafka/utils.go [hunk 2]
```diff
@@ -96,3 +97,13 @@ func makeBlockGroupOutput(blockchain *blockchain.BlockChain, block *types.Block,
 	r["transactions"] = rpcTransactions
 	return r
 }
+
+func intToBytes(num uint64) []byte {
+	buffer := make([]byte, 8)
+	binary.BigEndian.PutUint64(buffer, num)
+	return buffer
+}
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/kafka/utils.go:105]: Why don't you use `common.Int64ToByteBigEndian`?

File: datasync/chaindatafetcher/kafka/config.go (hunk 1 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/config.go:43]: Right, but the default value of kafka configuration is 1000000.

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/config.go:43]: This value may be modified. The smaller segment size would be better for the performance.

File: datasync/chaindatafetcher/kafka/utils.go (hunk 2 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/utils.go]: Reflected. a990b9f

**Comment by jeongkyun-oh** [comment]
> I have some questions and comments.
> 
> 1. Why splitting is required in Kafka? What are the benefits?
> 2. I think publishing segments in parallel can enhance speed.
> 3. A data is sent by calling `publish()` one time, and is received by calling several `subscribe()`. How do we know if the received data is a splitted one? After receiving, the segments need to be combined into one.
>    It would be nice to specify a protocol including data ID and segment number. Also, one `publish()` call should be correspond to one `subscribe()`.

@winnie-byun Thanks for your comments.
1. I've got the feedback that the large message would affect Kafka's performance badly. The [benchmark](https://engineering.linkedin.com/kafka/benchmarking-apache-kafka-2-million-writes-second-three-cheap-machines) said that the throughput would be worse if the larger messages are published logarithmically.
2. The order would be important, so I think parallelism is not a good option.
3. Yes, you're right. This PR includes only the producer part. The consumer part will be updated soon. The header includes all the information for resembling data (segment index, total segments, checksum). Do you think this is not enough? I will commit the data ID part as well in this PR. Also, `Publish` corresponds to  `Consume`. `Subscribe` keeps consuming in a loop. What do you think?

File: cmd/utils/flags.go [hunk 3]
```diff
@@ -786,6 +786,11 @@ var (
 		Usage: "The max size of a message produced by Kafka producer ",
 		Value: kafka.DefaultMaxMessageBytes,
 	}
+	ChainDataFetcherKafkaSegmentSizeFlag = cli.IntFlag{
+		Name:  "chaindatafetcher.kafka.segment.size",
+		Usage: "The kafka data segment size",
```
Comment (KimKyungup) [inline cmd/utils/flags.go:791]:
```suggestion
		Usage: "The kafka data segment size(in MB)",
```

**Review by winnie-byun — DISMISSED** [review]
Thanks for the kind explaination. I didn't see `makeProducerMessage` before.
This code looks good to be :)

File: datasync/chaindatafetcher/kafka/config.go (hunk 1 above)
Comment (KimKyungup) [reply datasync/chaindatafetcher/kafka/config.go:43]: MiB is 1024^2. MB is 1000^2.

File: datasync/chaindatafetcher/kafka/config.go [hunk 4]
```diff
@@ -66,6 +68,7 @@ func GetDefaultKafkaConfig() *KafkaConfig {
 		TopicResourceName:    DefaultTopicResourceName,
 		Partitions:           DefaultPartitions,
 		Replicas:             DefaultReplicas,
+		SegmentSize:          DefaultSegmentSize,
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/kafka/config.go:71]: How about this?

```suggestion
		SegmentSizeByte:          DefaultSegmentSizeByte,
```

File: datasync/chaindatafetcher/kafka/kafka.go [hunk 5]
```diff
@@ -76,17 +90,53 @@ func (k *Kafka) ListTopics() (map[string]sarama.TopicDetail, error) {
...
 	if err != nil {
 		return err
 	}
-
-	item := &sarama.ProducerMessage{
-		Topic: topic,
-		Value: sarama.StringEncoder(data),
+	segments, totalSegments := k.split(data)
+	for idx, segment := range segments {
+		item := k.makeProducerMessage(topic, segment, uint64(idx), uint64(totalSegments))
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/kafka/kafka.go:133]:
```suggestion
		msg := k.makeProducerMessage(topic, segment, uint64(idx), uint64(totalSegments))
```

File: datasync/chaindatafetcher/kafka/kafka.go [hunk 6]
```diff
@@ -76,17 +90,53 @@ func (k *Kafka) ListTopics() (map[string]sarama.TopicDetail, error) {
...
+			{
+				Key:   []byte(KeyCheckSum),
+				Value: checkSum[:],
+			},
+		},
+		Value: sarama.ByteEncoder(segment),
+	}
+}
+
 func (k *Kafka) Publish(topic string, msg interface{}) error {
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/kafka/kafka.go:126]:
```suggestion
func (k *Kafka) Publish(topic string, data interface{}) error {
```

File: datasync/chaindatafetcher/kafka/kafka.go [hunk 7]
```diff
@@ -76,17 +90,53 @@ func (k *Kafka) ListTopics() (map[string]sarama.TopicDetail, error) {
...
+	for len(data) > size {
+		segments = append(segments, data[:size])
+		data = data[size:]
+	}
+	segments = append(segments, data)
+	return segments, len(segments)
+}
+
+func (k *Kafka) makeProducerMessage(topic string, segment []byte, segmentIdx, totalSegments uint64) *sarama.ProducerMessage {
+	checkSum := md5.Sum(segment)
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/kafka/kafka.go:105]: Let's remove the checkout as we reviewed

**Review by KimKyungup — COMMENTED** [review]
LGTM except minor comments

File: datasync/chaindatafetcher/kafka/config.go [hunk 8]
```diff
@@ -40,6 +40,7 @@ const (
 	DefaultTopicResourceName    = "en-0"
 	DefaultMaxMessageBytes      = 1000000
 	DefaultRequiredAcks         = 1
+	DefaultSegmentSizeBytes     = 1000000 // 1 MB
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/kafka/config.go:43]:
```suggestion
	DefaultSegmentSizeBytes     = 1000000 // 1 MiB
```

**Review by KimKyungup — APPROVED** [review]

File: cmd/utils/flags.go (hunk 3 above)
Comment (jeongkyun-oh) [reply cmd/utils/flags.go]: Updated. d471e8e2884ae984a4c0ef3daeb5a92546f2f2a3

File: datasync/chaindatafetcher/kafka/config.go (hunk 4 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/config.go]: Updated with `SegmentSizeBytes` in plural. d471e8e2884ae984a4c0ef3daeb5a92546f2f2a3

File: datasync/chaindatafetcher/kafka/kafka.go (hunk 5 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka.go]: Reflected. d471e8e2884ae984a4c0ef3daeb5a92546f2f2a3

File: datasync/chaindatafetcher/kafka/kafka.go (hunk 6 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka.go]: Reflected. d471e8e2884ae984a4c0ef3daeb5a92546f2f2a3

File: datasync/chaindatafetcher/kafka/kafka.go (hunk 7 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka.go]: Reflected. e28ea0b

**Review by aidan-kwon — APPROVED** [review]

**Review by yoomee1313 — APPROVED** [review]

**Review by winnie-byun — APPROVED** [review]

**Review by ehnuje — APPROVED** [review]

**Pull request merged**